### PR TITLE
Emscripten fetch

### PR DESCRIPTION
**Pull request by juj**
Emscripten Fetch API provides multithreaded XHR + IndexedDB access in async, sync and waitable modes.

Adding this as a PR already in case anyone is interested in reading through a diff. Not yet for merge.

### CODE REVIEW

**Comment by kripken** [comment]
Very nice!

Overall the API looks good to me.

How will this cooperate with existing JS FS stuff? Can MEMFS, NODEFS, IDBFS and others still work? And can they still work in the same build?

**Comment by kripken** [comment]
Another issue to think about is that [BrowserFS](https://github.com/jvilk/BrowserFS) integrates with our filesystem. It provides very useful functionality, and is constantly adding more (e.g. it's used by the Internet Archive emulation projects using emscripten). We should make sure changes on our side don't cause problems for that integration. Also perhaps there are new integration opportunities with these changes. cc @jvilk

**Comment by jvilk** [comment]
@kripken Theoretically speaking, if `NODEFS` works without changes, BrowserFS should continue to work. But thanks for pinging me; I can test the changes once they are complete if you think it should still work.

What sort of new integration opportunities do you envision? Technically, anything you can do to further integrate the Node API is applicable to BrowserFS (with the exception of streams, which we currently do not support). e.g. BrowserFS supports multiple outstanding FS requests, so it should work in a multithreaded environment where all requests are multiplexed onto the context hosting the file system.

**Comment by kripken** [comment]
@jvilk: cool, good to know if we keep `NODEFS` working then we should be in good shape.

I didn't have anything else specific in mind here, just that @juj is exploring new filesystem implementation options, and working towards multithreaded filesystem support and other things on top of it (see [this post](https://groups.google.com/forum/#!topic/emscripten-discuss/bh3674QMyeY)). And BrowserFS is another major project in the filesystem space, so maybe a good time to discuss things and see if there are interesting options.

**Comment by juj** [comment]
> How will this cooperate with existing JS FS stuff? Can MEMFS, NODEFS, IDBFS and others still work? And can they still work in the same build?

The emscripten_fetch() API does not currently interact with MEMFS or other filesystem types at all. emscripten_fetch() requests either load files directly to memory as byte buffers, or to IndexedDB, under a storage database that emscripten_fetch() creates for itself. So MEMFS and other filesystems can be used in a same program that also calls to emscripten_fetch(), but there is no automated connection and emscripten_fetch() is not aware of the filesystem, except via the namespace of files in IndexedDB.

I could make an automation where emscripten_fetch() would fetch data to MEMFS in addition to byte buffers in memory, but that might create an awkward-looking inversion of data flow: in the ASMFS filesystem, ASMFS is built on top of emscripten_fetch(), i.e. ASMFS does emscripten_fetch() requests to load and store files from CDNs/IndexedDB. emscripten_fetch() does not know about a filesystem.

If emscripten_fetch() was made to store to MEMFS, it might be peculiar, because then the data flow would be the opposite and go from emscripten_fetch() -> filesystem, when in ASMFS mode, it is the filesystem that drives emscripten_fetch()es.

It would be doable, but not putting much attention to that at least initially, since it's also doable to fopen+fwrite+fclose the results of an emscripten_fetch() to a file in MEMFS manually.

**Comment by juj** [comment]
ASMFS on the other hand is a complete replacement of MEMFS and all the other mount-based filesystems, and not intended to be used simultaneously. When ASMFS is used, one can pair the build with `-s NO_FILESYSTEM=1`, since ASMFS implements the syscalls in C/C++ land, along with stdout streams and so forth. ASMFS manages a filesystem inode structure in memory, and builds on top of the filesystem namespace that emscripten_fetch() has established on top of IndexedDB.

**Comment by kripken** [comment]
Sounds good about fetch and MEMFS being orthogonal.

I worry a little about ASMFS not being compatible with the others. The name might be confusing since the other *FSes are mutually compatible. Also, `NO_FILESYSTEM` not disabling it seems surprising as well. Maybe this is just a naming and documentation issue though.

**Comment by jvilk** [comment]
> ASMFS manages a filesystem inode structure in memory, and builds on top of the filesystem namespace that emscripten_fetch() has established on top of IndexedDB.

Is there room for plugins to ASMFS that manage their own inodes?

**Comment by juj** [comment]
> I worry a little about ASMFS not being compatible with the others. The name might be confusing since the other *FSes are mutually compatible. Also, NO_FILESYSTEM not disabling it seems surprising as well. Maybe this is just a naming and documentation issue though.

In my current test builds one passes `-s ASMFS=1 -s NO_FILESYSTEM=1` to build. Perhaps should look at some other name, agreed there. It is very difficult to try to make ASMFS compatible with the others, since it would go back to proxying land immediately and sequentialize FS operations, which ASMFS intends to remove.

> Is there room for plugins to ASMFS that manage their own inodes?

Not looking at that initially, but I think that's a good feature to have once this matures a bit. There will be an inode type field that allows defining what type of item each inode is, that is probably like the way to expand later on.

File: tests/fetch/cached_xhr.cpp [hunk 1]
```diff
@@ -0,0 +1,56 @@
+#include <string.h>
```
Comment (asRIA) [inline tests/fetch/cached_xhr.cpp]: Name of the doesn't follow convention `example_...` :)

File: tests/fetch/example_idb_delete.cpp [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+void failure(emscripten_fetch_t *fetch)
+{
+  printf("Deleting file from IDB failed.\n");
+  emscripten_fetch_close(fetch);
+}
+
+int main()
+{
+  emscripten_fetch_attr_init(&attr);
```
Comment (asRIA) [inline tests/fetch/example_idb_delete.cpp]: `attr` is never created

File: system/include/emscripten/fetch.h [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+// Synchronously blocks to wait for the given fetch operation to complete. This operation is not allowed in the main browser
+// thread, in which case it will return EMSCRIPTEN_RESULT_NOT_SUPPORTED. Pass timeoutMSecs=infinite to wait indefinitely. If
+// the wait times out, the return value will be EMSCRIPTEN_RESULT_TIMEOUT.
+// The onsuccess()/onerror()/onprogress() handlers will be called in the calling thread from within this function before
+// this function returns.
+EMSCRIPTEN_RESULT emscripten_fetch_wait(emscripten_fetch_t *fetch, double timeoutMSecs);
+
+// Closes a finished or an executing fetch operation and frees up all memory. If the fetch operation was still executing, the
+// onerror() handler will be called in the calling thread before this function returns.
+EMSCRIPTEN_RESULT emscripten_fetch_close(emscripten_fetch_t *fetch);
```
Comment (asRIA) [inline system/include/emscripten/fetch.h]: Comment might be misleading, because it will not free `requestData`

File: tests/fetch/example_idb_store.cpp [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+  emscripten_fetch_close(fetch);
+}
+
+void persistFileToIndexedDB(const char *outputFilename, uint8_t *data, size_t numBytes)
+{
+  emscripten_fetch_attr_t attr;
+  emscripten_fetch_attr_init(&attr);
+  strcpy(attr.requestMethod, "EM_IDB_STORE");
+  attr.attributes = EMSCRIPTEN_FETCH_REPLACE | EMSCRIPTEN_FETCH_PERSIST_FILE;
+  attr.requestData = (char *)data;
```
Comment (asRIA) [inline tests/fetch/example_idb_store.cpp]: Data from `requestData` is copied by `HEAP8.slice`. 
`emscripten_fetch_close` frees the memory shallowly, hence allocated memort in line 35 will leak. 
It's not a big deal here, but it's not documented who owns memory of `requestData` and if `emscripten_fetch_close` will free it. The same applies for code blocks for the documentation.

**Comment by gouletr** [comment]
Awesome work! Can't wait to try it out. 👍

File: src/Fetch.js [hunk 5]
```diff
@@ -0,0 +1,530 @@
...
+  var fetchAttrNoDownload = !!(fetchAttributes & 32/*EMSCRIPTEN_FETCH_NO_DOWNLOAD*/);
+  var fetchAttrSynchronous = !!(fetchAttributes & 64/*EMSCRIPTEN_FETCH_SYNCHRONOUS*/);
+  var fetchAttrWaitable = !!(fetchAttributes & 128/*EMSCRIPTEN_FETCH_WAITABLE*/);
+
+  var userNameStr = userName ? Pointer_stringify(userName) : undefined;
+  var passwordStr = password ? Pointer_stringify(password) : undefined;
+  var overriddenMimeTypeStr = overriddenMimeType ? Pointer_stringify(overriddenMimeType) : undefined;
+
+  var xhr = new XMLHttpRequest();
+  xhr.timeout = timeoutMsecs;
```
Comment (juj) [inline src/Fetch.js]: Thanks, very precise finding. Fixed.

**Comment by juj** [comment]
> I've seen that status code is verified only if it's 200, but other codes like 304.

The 304 codes cannot currently happen because they would require `If-Modified-Since` fields to be present. Once the usage of these will be added to Fetch, the handling of 304 will be added as well. (The code will need to also handle the absence of byte data on HTTP 304, and depending on the Fetch request flags, the result will need to be handled a bit different)

**Comment by juj** [comment]
> That means instead of being reactive for callbacks that may happen in any time, we call poll method that releases all callbacks aggregated in meantime. 

It is possible for an user of the Fetch API to either use events, or poll, but I'm not sure if removing callback functions after first adding them will be a supported feature. Right now that looks like a clean thing that works (one can set the callback function pointer to 0 mid-fly of an async fetch), but I'll reserve a chance to change my mind if big issues come up with that. At worst the user code can use the `userData` pointer to detect if the completion has been handled or not via a poll or an event callback, to avoid handling twice, if it's also polling.

File: src/Fetch.js [hunk 6]
```diff
@@ -0,0 +1,530 @@
...
+  xhr.withCredentials = withCredentials;
+#if FETCH_DEBUG
+  console.log('fetch: xhr.timeout: ' + xhr.timeout + ', xhr.withCredentials: ' + xhr.withCredentials);
+  console.log('fetch: xhr.open(requestMethod="' + requestMethod + '", url: "' + url_ +'", userName: ' + userNameStr + ', password: ' + passwordStr + ');');
+#endif
+  xhr.open(requestMethod, url_, !fetchAttrSynchronous, userNameStr, passwordStr);
```
Comment (juj) [inline src/Fetch.js]: Thanks, this is documented in https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/timeout, adapted for that.

File: system/include/emscripten/fetch.h [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+	// Specifies the length of the above data block in bytes. When the download finishes, this field will be valid even if
+	// EMSCRIPTEN_FETCH_LOAD_TO_MEMORY was not specified.
+	uint64_t numBytes; // XXX 16
+
+	// If EMSCRIPTEN_FETCH_STREAM_DATA is being performed, this indicates the byte offset from the start of the stream
+	// that the data block specifies. (for onprogress() streaming XHR transfer, the number of bytes downloaded so far before this chunk)
+	uint64_t dataOffset; // XXX 24
+
+	// Specifies the total number of bytes that the response body will be.
+	uint64_t totalBytes; // XXX 32
```
Comment (juj) [inline system/include/emscripten/fetch.h]: Yeah, that's true. Added a note about that.

File: system/include/emscripten/fetch.h (hunk 3 above)
Comment (juj) [inline system/include/emscripten/fetch.h]: That was currently unimplemented, `emscripten_fetch_close` will free `data` and call `onerror` if the close caused a fetch to abort mid-flight, so the intent is that the function will behave like documented.

File: tests/fetch/cached_xhr.cpp (hunk 1 above)
Comment (juj) [inline tests/fetch/cached_xhr.cpp]: This particular file, unlike the others, is used as a test in the suite, so didn't have the prefix `example_` there. The ones with `example_` are intended to be compact and unlike the tests they don't verify that the implementation is correct (those lines bloat up the tests by quite a bit).

File: tests/fetch/cached_xhr.cpp [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+    assert(fetch->numBytes == fetch->totalBytes);
+    assert(fetch->data != 0);
+    printf("Finished downloading %llu bytes\n", fetch->numBytes);
+    emscripten_fetch_close(fetch);
+
+    // Test that the file now exists:
+    fetchFromIndexedDB();
+  };
+  attr.onprogress = [](emscripten_fetch_t *fetch) {
+    printf("Downloading.. %.2f%% complete.\n", (fetch->dataOffset + fetch->numBytes) * 100.0 / fetch->totalBytes);
```
Comment (juj) [inline tests/fetch/cached_xhr.cpp]: In the test harness the `Content-Length` fields are always present, so `fetch->totalBytes` is always there, but agreed that for people reading the code to use as copy-paste examples to their project, it's an omission. Added the checks to the test suite so that people won't get any surprises.

File: tests/fetch/example_idb_delete.cpp (hunk 2 above)
Comment (juj) [inline tests/fetch/example_idb_delete.cpp]: Thanks, fixed.

File: tests/fetch/example_idb_store.cpp (hunk 4 above)
Comment (juj) [inline tests/fetch/example_idb_store.cpp]: Yeah, the implementation of `emscripten_fetch_close()` was partial, this is now implemented, and also documented more explicitly.

File: src/library_html5.js [hunk 9]
```diff
@@ -1557,7 +1557,7 @@ var LibraryJSEvents = {
   },
 
   emscripten_set_pointerlockchange_callback: function(target, userData, useCapture, callbackfunc) {
-    if (!document.body.requestPointerLock && !document.body.mozRequestPointerLock && !document.body.webkitRequestPointerLock && !document.body.msRequestPointerLock) {
+    if (!document.body || (!document.body.requestPointerLock && !document.body.mozRequestPointerLock && !document.body.webkitRequestPointerLock && !document.body.msRequestPointerLock)) {
```
Comment (kripken) [inline src/library_html5.js]: unrelated changes?

File: src/library_syscall.js [hunk 10]
```diff
@@ -969,7 +969,7 @@ var SyscallsLibrary = {
       stream.getdents = FS.readdir(stream.path);
     }
     var pos = 0;
-    while (stream.getdents.length > 0 && pos + {{{ C_STRUCTS.dirent.__size__ }}} < count) {
```
Comment (kripken) [inline src/library_syscall.js]: was this a bug? could we add a test?

File: emcc.py [hunk 11]
```diff
@@ -1692,6 +1707,35 @@ def repl(m):
       if shared.Settings.USE_PTHREADS:
         shutil.copyfile(shared.path_from_root('src', 'pthread-main.js'), os.path.join(os.path.dirname(os.path.abspath(target)), 'pthread-main.js'))
 
+      # Generate the fetch-worker.js script for multithreaded emscripten_fetch() support if targeting pthreads.
+      if shared.Settings.FETCH and shared.Settings.USE_PTHREADS:
+        src = open(final, 'r').read()
```
Comment (kripken) [inline emcc.py]: maybe move this to a function `make_fetch_worker` in `shared.py`?

File: src/library_syscall.js (hunk 10 above)
Comment (juj) [inline src/library_syscall.js]: Added a test, and moved this bit to a separate PR #4779.

File: emcc.py (hunk 11 above)
Comment (juj) [inline emcc.py]: Yeah, that's a good idea. Moved.

**Comment by juj** [comment]
Ok, now adjusted to the review, and spliced off separate features to their individual PRs. @kripken: how does this look now?

**Comment by kripken** [comment]
lgtm, we can do the rest in followups.

**Comment by juj** [comment]
Ran all test suites and each look clean, so this should be safe to merge in. This PR needs a new tag though to reset caches, I think we should add one right before and after.

**Comment by juj** [comment]
Ok, now that the merge is done, I'll push this in as well, and tag around this.
